packages/notebook/src/actions.tsx

```diff
@@ -2273,51 +2273,67 @@ Please wait for the complete rendering before invoking that action.`,
             });
             return Promise.resolve(false);
           }
+          let promise = Promise.resolve();
           if (sessionContext.hasNoKernel) {
-            void sessionContextDialogs.selectKernel(sessionContext);
-            return Promise.resolve(false);
-          }
-          const deletedCells = notebook.model?.deletedCells ?? [];
-          executionScheduled.emit({ notebook, cell });
-          return CodeCell.execute(cell as CodeCell, sessionContext, {
-            deletedCells,
-            recordTiming: notebook.notebookConfig.recordTiming
-          })
-            .then(reply => {
-              deletedCells.splice(0, deletedCells.length);
-              if (cell.isDisposed) {
-                return false;
+            promise = sessionContext.startKernel().then(shouldSelect => {
+              if (shouldSelect) {
+                return sessionContextDialogs.selectKernel(sessionContext);
               }
+            });
+          }
+          return promise.then(() => {
+            if (sessionContext.hasNoKernel) {
+              // Session has still no kernel, so we can't execute the cell.
+              return false;
+            }
 
-              if (!reply) {
-                return true;
-              }
-              if (reply.content.status === 'ok') {
-                const content = reply.content;
+            let deletedCells = notebook.model?.deletedCells ?? [];
```

Comment:
code below is just wrapping it inside promise. Git diff is a bit confusing, but I do not edit details.